frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -994,11 +995,16 @@ func dispatchRun(d *dispatchState, c *instructions.RunCommand, proxy *llb.ProxyE
 }
 
 func dispatchWorkdir(d *dispatchState, c *instructions.WorkdirCommand, commit bool, opt *dispatchOpt) error {
-	d.state = d.state.Dir(c.Path)
-	wd := c.Path
-	if !path.IsAbs(c.Path) {
-		wd = path.Join("/", d.image.Config.WorkingDir, wd)
+	platformOS := runtime.GOOS
+	if d != nil && d.platform != nil {
```

Comment:
I don't think these can be `nil`?

Reply:
>  d.platform can be, 

What case is that? https://github.com/moby/buildkit/blob/master/frontend/dockerfile/dockerfile2llb/convert.go#L453

Reply:
Will check, but the reason I added this was because at one point I got a nil pointer dereference. Otherwise I would not have touched it. It's a good idea to check pointers before we try dereferencing, but if you prefer, I can remove this check. Either way is fine with me.

Reply:
Removed the check

Reply:
This is the reason for this check:

https://github.com/moby/buildkit/actions/runs/5158872514/jobs/9293127738?pr=3908#step:7:1692

Reply:
I'm mostly worried that if this is nil there might be a better place to set it or maybe it is nil because of an issue.

Reply:
I've been thinking about our discussion here in regards to checking for nil pointers and the discussion bellow about normalizing a path that may already be normalized in the client before sending it (details in that thread).

In general, I prefer to check all inputs as close as possible to where I need to use them. The exception to this is when I need to check user input, from say an API call, in which case the payload should of course be validated.

The reason for this is that we cannot trust the caller of a function. Bugs happen. Mistakes happen. When those mistakes happen, it's important to either be flexible and attempt to sanitize the input if possible (and log a warning), or return a meaningful error back to the caller. In my experience, pointers should always be checked before they are accessed. 

If we end up de-referencing a nil pointer out of the belief that there is no way a value can be nil, we end up panicking the code, which leads to the ```buildkit``` daemon dying (unless we ```recover()```). I believe it's far more valuable to check for a nil pointer and if the expectation is that it should not be nil, to return a meaningful error.

That error bubbles up to the caller and eventually reaches the user. A properly wrapped error can help us debug any issues without killing the daemon, and allows us to potentially recover from the error a lot easier.

In short, the correctness of the behavior of one function should not depend on the assumption that the input provided to it by the caller, is correct. Every input value that a particular function plans to use, should be checked.

I am not properly familiar with the design decisions involved here. I am unsure if ```d.platform``` being ```nil``` here is by design or by accident, but I believe that ```dispatchWorkdir()``` should return an error if any value it depends on is unexpected. A panic is less valuable.

This is just my 2 cents. In the end, I will adhere to any design decision you folks prefer.